javaslang/src/main/java/javaslang/collection/AbstractMultimap.java

```diff
@@ -0,0 +1,455 @@
+/*     / \____  _    _  ____   ______  / \ ____  __    _______
+ *    /  /    \/ \  / \/    \ /  /\__\/  //    \/  \  //  /\__\   JΛVΛSLΛNG
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/ \ /__\ \   Copyright 2014-2016 Javaslang, http://javaslang.io
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+import javaslang.API;
+import javaslang.Lazy;
+import javaslang.Tuple;
+import javaslang.Tuple2;
+import javaslang.control.Option;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.function.*;
+
+/**
+ * An {@link Multimap} implementation (not intended to be public).
+ *
+ * @param <K> Key type
+ * @param <V> Value type
+ * @param <M> Multimap type
+ * @author Ruslan Sennov
+ * @since 2.0.0
+ */
+ abstract class AbstractMultimap<K, V, M extends Multimap<K, V>> implements Multimap<K, V> {
+
+    interface SerializableSupplier<T> extends Supplier<T>, Serializable {
+    }
+
+    private static final long serialVersionUID = 1L;
+
+    final Map<K, Traversable<V>> back;
+    final SerializableSupplier<Traversable<?>> emptyContainer;
+    private final ContainerType containerType;
+
+    AbstractMultimap(Map<K, Traversable<V>> back, ContainerType containerType, SerializableSupplier<Traversable<?>> emptyContainer) {
+        this.back = back;
+        this.containerType = containerType;
+        this.emptyContainer = emptyContainer;
+    }
+
+    abstract <K2, V2> Map<K2, V2> emptyMapSupplier();
+
+    abstract <K2, V2> Multimap<K2, V2> emptyInstance();
+
+    abstract <K2, V2> Multimap<K2, V2> createFromMap(Map<K2, Traversable<V2>> back);
+
+    @SuppressWarnings("unchecked")
+    private <K2, V2> Multimap<K2, V2> createFromEntries(Iterable<? extends Tuple2<? extends K2, ? extends V2>> entries) {
```

Comment:
I think it can be `Tuple2` instead of `? extends Tuple2`.

For K2 and V2 I'm not sure but probably we do not need the `? extends` because this method is private.